src/plugins/oidc/login.ts

```diff
@@ -57,70 +44,142 @@ const authorize = async (org: OrgData, scope: string) => {
  * authorization server will return "authorization_pending", in which case this
  * function will return undefined.
  */
-const fetchOidcToken = async (org: OrgData, authorize: AuthorizeResponse) => {
-  if (org.providerType === undefined) {
-    throw "Login requires a configured provider type.";
-  }
-  const init = {
-    method: "POST",
-    headers: OIDC_HEADERS,
-    body: urlEncode({
-      client_id: org.clientId,
-      device_code: authorize.device_code,
-      grant_type: DEVICE_GRANT_TYPE,
-    }),
-  };
-  validateProviderDomain(org);
-  const url =
-    org.providerType === "okta"
-      ? `https:${org.providerDomain}/oauth2/v1/token`
-      : org.providerType === "ping"
-        ? `https://${org.providerDomain}/${org.environmentId}/as/token`
-        : throwAssertNever(org.providerType);
+export const fetchOidcToken = async <T>(request: {
+  url: string;
+  init: RequestInit;
+}) => {
+  const { url, init } = request;
   const response = await fetch(url, init);
-
   if (!response.ok) {
     if (response.status === 400) {
       const data = await response.json();
       if (data.error === "authorization_pending") return undefined;
     }
     await validateResponse(response);
   }
-  return (await response.json()) as TokenResponse;
+  return (await response.json()) as T;
+};
+
+const providerType: (org: OrgData) => NonNullable<string> = (org) => {
+  if (org.providerType === undefined) {
+    throw "Login requires a configured provider type.";
+  }
+  return org.providerType;
 };
 
 /** Waits until user device authorization is complete
  *
  * Returns the OIDC token after completion.
  */
-const waitForActivation = async (
-  org: OrgData,
-  authorize: AuthorizeResponse
+export const waitForActivation = async <A, T>(
+  authorize: A,
+  extractExpiryInterval: (authorize: A) => {
+    expires_in: number;
+    interval: number;
+  }, // Aws implementation differs from standard OIDC response, need function to extract expiry
+  tokenRequest: { url: string; init: RequestInit }
 ) => {
   const start = Date.now();
-  while (Date.now() - start <= authorize.expires_in * 1e3) {
-    const response = await fetchOidcToken(org, authorize);
-    if (!response) await sleep(authorize.interval * 1e3);
+  const { expires_in, interval } = extractExpiryInterval(authorize);
+  while (Date.now() - start <= expires_in * 1e3) {
+    const response = await fetchOidcToken<T>(tokenRequest);
+    if (!response) await sleep(interval * 1e3);
     else return response;
   }
   throw "Expired awaiting in-browser authorization.";
 };
 
+export const oidcLoginSteps = (
+  org: OrgData,
+  scope: string,
+  urls: () => { deviceAuthorizationUrl: string; tokenUrl: string }
+) => {
+  const { deviceAuthorizationUrl, tokenUrl } = urls();
+
+  const buildOidcAuthorizeRequest = () => {
+    if (org.providerType === undefined) {
+      throw "Login requires a configured provider type.";
```

Comment:
What is the user supposed to do with this? Also, do they know what a "provider type" is?

We should direct the user to contact their P0 admin:

```suggestion
      throw "Your organization's login configuration does not support this access. Your P0 admin will need to install a supported OIDC provider in order for you to use this command.";
```